### ISSUE TITLE
(from email) Some questions about example code for X11rb

### ISSUE THREAD

**Issue by eduardosm** [psychon/x11rb#514]
I have received the following email, which I think it is best discussed here:

> I am learning X windows system to write a window manager in Rust recently, and I noticed your repo on Github.
> After reading the example "simple_window_manager.rs" in the example code of x11rb,
> I found the way of handling event loop in the example is quite different in other WM (dwm, katriawm).
> To my knowledge, conn.wait_for_event() is the blocking way, so it will wait for a new event; and then enter a loop until the conn.poll_for_event() returns None since it is non-blocking, and process all Expose events.
> So I am writing to inquire about what is the advantage of using a non-blocking way and processing the Expose events at once.
>
> Attached is the event loop code of x11rb example and dwm.
>
> ![x11rb](https://user-images.githubusercontent.com/761151/89104729-c4d94400-d41b-11ea-8697-84a15ad6590f.png)
>
> ![dwm](https://user-images.githubusercontent.com/761151/89104730-c73b9e00-d41b-11ea-974b-62b13d7fdfa8.png)

**Comment by eduardosm** [comment]
>  So I am writing to inquire about what is the advantage of using a non-blocking way and processing the Expose events at once.

I believe it is to avoid handling redundant Expose events. @psychon actually wrote this example, maybe he can give a better explanation.

**Comment by psychon** [comment]
Heh. I got the same mail. :-)

Yesterday, I answered the following:

-----

the benefit is that it might save redraws. When a non-rectangular area
is exposed, the X11 server generates multiple expose events. So, for
example, if an "L"-shaped area is exposed, one would get two expose
events for the two rectangles that the "L" is composed of.

dwm seems to redraw on each expose event (*). So, in the above
situation, it would draw its bar two times (*).

The sample_window_manager just remembers "this needs to be redrawn".
When it handled all pending events, it goes through the list of "this
needs to be redrawn" and handles all of them once.

(*) However, the above is actually wrong. Here is the function handling
expose() events in dwm:

void
expose(XEvent *e) {
	Monitor *m;
	XExposeEvent *ev = &e->xexpose;

	if(ev->count == 0 && (m = wintomon(ev->window)))
		drawbar(m);
}

The important part is "ev->count == 0". From
https://www.x.org/releases/X11R7.6/doc/xproto/x11protocol.html#events:Expose:

> If count is zero, then no more Expose events for this window follow.> If count is nonzero, then at least that many more Expose events for>
this window follow (and possibly more).

So, with this simple check, dwm also manages to only redraw once.

Still, I feel like the approach from simple_window_manager is slightly
more general and might save more redraws in random edge cases. But in
the end, I guess it does not really matter and either approach is fine.

I hope this answers your question.

Cheers,
Uli
